client/client-lib/src/mint/mod.rs

```diff
@@ -346,19 +346,15 @@ impl MintClient {
         )
     }
 
-    async fn new_note_secret(&self, amount: Amount) -> DerivableSecret {
-        let mut new_idx;
-        loop {
-            let mut dbtx = self.start_dbtx().await;
-            new_idx = self.get_next_note_index(&mut dbtx, amount).await;
-            dbtx.insert_entry(&NextECashNoteIndexKey(amount), &new_idx.next().as_u64())
-                .await
-                .expect("DB error");
-            if dbtx.commit_tx().await.is_ok() {
-                break;
-            }
-        }
-
+    async fn new_note_secret(
+        &self,
+        amount: Amount,
+        dbtx: &mut DatabaseTransaction<'_>,
+    ) -> DerivableSecret {
+        let new_idx = self.get_next_note_index(dbtx, amount).await;
+        dbtx.insert_entry(&NextECashNoteIndexKey(amount), &new_idx.next().as_u64())
```

Comment:
So as we can see here, I needed to modify this so that it takes an existing db transaction instead of creating a new one. This fix still isn't complete, because the insert_entry here will fail when multiple threads try to generate a new note secret. I think the "proper" fix is to propagate the error up to higher layers that can re-try. Rocksdb will fail at commit time (instead of here), but in both cases the operation should be retried.

This fix is currently blocking me from changing the test_parallel_issuances test to use sqlite. But in my opinion, that change is a bit outside of the scope of this PR as it is also making a functional change to the client.

I know @elsirion is still planning on redoing a lot of the client to use state machines. So perhaps messing with this too much is not necessary.

Reply:
What makes this change necessary? Does SQLite have different different consistency guarantees than we expect? Is it the difference between optimistic vs. pessimistic tx mentioned further down in the tests?

Reply:
Yes, it because sqlite uses locks to implement transactions (pessmisitic). Sqlite will return "db busy" error when there's concurrent write transactions (since it needs to lock the WAL to insert into the db). Here we were creating a new write tx when we already had one open, so I needed to re-use the existing transaction. 

The consistency guarantees are the same (both use snapshot isolation). It just a matter of when the transaction fails. Sqlite fails during the insert, whereas rocksdb fails during commit.

Reply:
Interesting, this points towards a much bigger problem: we currently assume that "DB error" is equal to "our disk is on fire" and thus `expect` them. From what I hear we now need to be able to propagate them to the function that created to DB tx so it can re-try?

Reply:
Yes exactly. We might not need to propagate it all the way back, I think it depends on the context of the insert. I figured that was a larger change out of scope for this PR, so I left it. Happy to take this on as well, just wanted to raise it in context of the client refactor you had started on.

Reply:
While I think it's better to have that dbtx passed there anyway, I'm confused why wouldn't a new tx here not work.

If another operation touches the counter, the insert will fail and we should retry ... oh.... but we are using `expect()` somewhere in the insert. I see.

Yeah, `expect`s on IO are not a good idea. :P

Reply:
We have an item for this already :) https://github.com/fedimint/fedimint/issues/374